update-engine/src/spec.rs

```diff
@@ -85,10 +85,93 @@ pub trait StepSpec: JsonSchema {
     type Error: AsError + fmt::Debug;
 }
 
+/// Represents a fully generic step specification, as can be serialized over
+/// JSON.
+///
+/// Since errors aren't directly serialized, they can be any type that
+/// implements [`AsError`].
+///
+/// Use this if you don't care about assigning types to any of the metadata
+/// components.
+pub struct GenericSpec<E> {
```

Comment:
This is a bit warty! I'm not happy about the fact that the E parameter exists here. The error absolutely does not matter for serialized data, just for running the engine itself.

I think I'm going to revisit this later. I think we may have to break this up into two separate types, one for serialized data and one for the error.